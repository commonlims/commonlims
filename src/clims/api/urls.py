from __future__ import absolute_import, print_function

from django.conf.urls import patterns
from django.conf import urls

from .endpoints.work_batch import WorkBatchEndpoint

from .endpoints.substance import SubstanceEndpoint
from .endpoints.substance_details import SubstanceDetailsEndpoint
from .endpoints.substance_property import SubstancePropertyEndpoint

from .endpoints.project import ProjectEndpoint
from .endpoints.project_property import ProjectPropertyEndpoint

from .endpoints.container import ContainerEndpoint
from .endpoints.container_property import ContainerPropertyEndpoint
from .endpoints.process_definitions import ProcessDefinitionsEndpoint

from .endpoints.user_files import UserFilesEndpoint

from .endpoints.work_batch_files import WorkBatchFilesEndpoint
from .endpoints.work_batch_file_details import WorkBatchFileDetailsEndpoint

from .endpoints.work_batch_settings import (WorkBatchSettingsEndpoint,
                                            WorkBatchSettingsDetailsEndpoint)

from .endpoints.work_batch_notes import WorkBatchNotesEndpoint
from .endpoints.work_batch_notes_details import WorkBatchNotesDetailsEndpoint

from .endpoints.plugin_actions import PluginActionsEndpoint

from .endpoints.workflow import WorkflowEndpoint
from .endpoints.task_definition import TaskDefinitionEndpoint

from .endpoints.substance_file import SubstanceFileEndpoint
from .endpoints.substance_file_details import (SubstanceFileDetailsEndpoint,
                                               SubstanceFileDemoDetailsEndpoint
                                               )

from .endpoints.organization_searches import OrganizationSearchesEndpoint

from .endpoints.process_assignments import ProcessAssignmentsEndpoint
from .endpoints.tasks import TasksEndpoint


def fmt(s):
    """Formats rules with common patterns"""
    s = s.replace('{org}', r'(?P<organization_slug>[^\/]+)')
    return s


def url(endpoint, *args, **kwargs):
    return urls.url(fmt(endpoint), *args, **kwargs)


urlpatterns = patterns(
    '',
    # Workflow
    url(r'^organizations/{org}/task-definitions/$',
        TaskDefinitionEndpoint.as_view(),
        name='clims-api-0-task-definitions'),
    url(r'^organizations/{org}/workflow/(?P<workflow_endpoint>[^\/]+)/$',
        WorkflowEndpoint.as_view(),
        name='clims-api-0-workflow-root'),
    url(r'^organizations/{org}/process-assignments/$',
        ProcessAssignmentsEndpoint.as_view(),
        name='clims-api-0-process-assignments'),
    url(r'^organizations/{org}/tasks/$',
        TasksEndpoint.as_view(),
        name='clims-api-0-tasks'),
    # Substances
    url(r'^substances/(?P<substance_id>[^\/]+)/$',
        SubstanceDetailsEndpoint.as_view(),
        name='clims-api-0-substance-details'),
    url(r'^organizations/{org}/substances/$',
        SubstanceEndpoint.as_view(),
        name='clims-api-0-substances'),
    url(r'^organizations/{org}/substances/property/(?P<prop>[^\/]+)/$',
        SubstancePropertyEndpoint.as_view(),
        name='clims-api-0-substance-property'),
    url(r'^organizations/{org}/containers/$',
        ContainerEndpoint.as_view(),
        name='clims-api-0-containers'),
    url(r'^organizations/{org}/containers/property/(?P<prop>[^\/]+)/$',
        ContainerPropertyEndpoint.as_view(),
        name='clims-api-0-container-property'),
    url(r'^organizations/{org}/projects/$',
        ProjectEndpoint.as_view(),
        name='clims-api-0-projects'),
    url(r'^organizations/{org}/projects/property/(?P<prop>[^\/]+)/$',
        ProjectPropertyEndpoint.as_view(),
        name='clims-api-0-project-property'),
    url(r'^organizations/{org}/substances/files/$',
        SubstanceFileEndpoint.as_view(),
        name='clims-api-0-organization-substances-files'),
    url(r'^organizations/{org}/substances/files/demo/$',
        SubstanceFileDemoDetailsEndpoint.as_view(),
        name='clims-api-0-organization-substances-files'),
    url(r'^organizations/{org}/substances/files/(?P<file_id>[^\/]+)/$',
        SubstanceFileDetailsEndpoint.as_view(),
        name='clims-api-0-organization-substances-file-details'),

    # work-batches: user task activities that have been grouped together in 1..n sized batches
    url(r'^organizations/{org}/work-batches/$',
        WorkBatchEndpoint.as_view(),
        name='clims-api-0-work-batches'),
    url(r'^work-batches/(?P<work_batch_id>[^\/]+)/files/$',
        WorkBatchFilesEndpoint.as_view(),
        name='clims-api-0-user-task-files'),
    url(r'^work-batches/(?P<work_batch_id>[^\/]+)/files/(?P<file_id>[^\/]+)/$',
        WorkBatchFileDetailsEndpoint.as_view(),
        name='clims-api-0-user-task-file-details'),
    url(r'^work-batches/(?P<work_batch_id>[^\/]+)/comments/$',
        WorkBatchNotesEndpoint.as_view(),
        name='clims-api-0-group-notes'),
    url(r'^work-batches/(?P<work_batch_id>[^\/]+)/comments/$(?P<note_id>[^\/]+)/',
        WorkBatchNotesDetailsEndpoint.as_view(),
        name='clims-api-0-group-notes-details'),
    url(r'^work-batch-settings/(?P<organization_slug>[^\/]+)/$',
        WorkBatchSettingsEndpoint.as_view(),
        name='clims-api-0-work-batch-settings'),
    url(r'^work-batch-settings/(?P<work_batch_type>[^\/]+)/$',
        WorkBatchSettingsDetailsEndpoint.as_view(),
        name='clims-api-0-work-batch-settings-details'),

    # Processes and Tasks
    url(r'^organizations/{org}/process-definitions/$',
        ProcessDefinitionsEndpoint.as_view(),
        name='clims-api-0-process-definitions'),
    url(r'^user-files/$',
        UserFilesEndpoint.as_view(),
        name='clims-api-0-user-files'),

    # Work batches:
    url(r'^organizations/{org}/work-batches/$',
        WorkBatchEndpoint.as_view(),
        name='clims-api-0-work-batches'),
    url(r'^user-tasks/(?P<user_task_id>[^\/]+)/files/$',
        WorkBatchFilesEndpoint.as_view(),
        name='clims-api-0-work-batch-files'),
    url(r'^work-batches/(?P<work_batch_id>[^\/]+)/files/(?P<file_id>[^\/]+)/$',
        WorkBatchFileDetailsEndpoint.as_view(),
        name='clims-api-0-work-batches-file-details'),

    # Plugins
    url(r'^plugins/{org}/(?P<plugin_id>[^\/]+)/actions/$',
        PluginActionsEndpoint.as_view(),
        name='clims-api-0-plugin-actions'),

    # Saved searches
    url(r'^organizations/{org}/saved-searches/$',
        OrganizationSearchesEndpoint.as_view(),
        name='clims-api-0-organization-saved-searches'),
)
