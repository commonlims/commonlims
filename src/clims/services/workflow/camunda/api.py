from __future__ import absolute_import
import requests
import logging
from six.moves.urllib.parse import urljoin
from string import Formatter

logger = logging.getLogger(__name__)

# Helper objects


class UnexpectedHttpResponse(Exception):
    pass


class Builder(object):
    """
    An object that creates the built object when called. This allows us to make it look
    like we're returning a factory when in fact we're returning an instance
    """

    def __init__(self, resource_type, url, many):
        self.resource_type = resource_type
        self.url = url
        self.many = many

    def __call__(self, **kwargs):
        return Requestor(self.resource_type, self.url, self.many, **kwargs)

    def __str__(self):
        return "<Builder {} at {}, many={}>".format(self.resource_type,
                                                    self.url, self.many)


class Requestor(object):
    """
    The object that knows what to do when requesting things. Returns a resource object or list
    of resource objects
    """

    def __init__(self, resource_type, url, many, **kwargs):
        self.url = url.format(**kwargs)
        self.validate_url()
        self.resource_type = resource_type
        self.many = many

    def _url(self, resource):
        return urljoin(self.base_url, resource)

    def validate_url(self):
        fmt = Formatter()
        if len(list(fmt.parse(self.url))) > 1:
            raise Exception("Uninitialized format specifiers in url {}".format(
                self.url))

    def get(self, **params):
        logger.debug("Fetching {} with params {}".format(self.url, params))
        response = requests.get(self.url, params=params)
        if response.status_code == 200:
            result = response.json()

            # TODO: Here we need to implement paging
            if not self.many:
                primary_key = result[self.resource_type.primary_key_field]
                return self.resource_type(self.url, result, primary_key)

            ret = list()
            for entry in result:
                primary_key = entry[self.resource_type.primary_key_field]
                ret.append(
                    self.resource_type(self.url + "/" + primary_key, entry,
                                       primary_key))
            return ret
        else:
            raise UnexpectedHttpResponse(self.url, response.status_code)

    def delete(self):
        response = requests.delete(self.url, params={"cascade": "true"})
        if response.status_code != 204:
            raise UnexpectedHttpResponse(self.url, response.status_code)

    def __str__(self):
        return "<{} url={}>".format(type(self).__name__, self.url)


class ResourceOfType(object):
    def __init__(self, resource_type, url):
        self.resource_type = resource_type
        self.url = url
        self.many = False

    def __get__(self, parent, type=None):
        url = parent.url + "/" + self.url
        return Builder(self.resource_type, url, many=self.many)


class ResourcesOfType(ResourceOfType):
    def __init__(self, resource_type, url):
        super(ResourcesOfType, self).__init__(resource_type, url)
        self.many = True


class CamundaResource(object):
    """Represents a single resource endpoint in the Camunda rest api"""

    primary_key_field = "id"

    def __init__(self, url, json, id):
        self.id = id
        self.url = url
        self.json = json

    def __str__(self):
        return "<{} url={}>".format(type(self).__name__, self.url)

    def __repr__(self):
        return self.url


# Resources


class ActivityInstance(CamundaResource):
    pass


class ProcessInstance(CamundaResource):
    pass

    # NOTE: This is a single entry, even though it's plural (it contains several intries)
    activity_instances = ResourceOfType(ActivityInstance, "activity-instances")


class ProcessDefinition(CamundaResource):
    pass


class Deployment(CamundaResource):
    pass


class Task(CamundaResource):
    pass


class CamundaApi(object):
    """Describes the Camunda API"""

    def __init__(self, url):
        if url.endswith("/"):
            url = url[:-1]
        self.url = url

    process_definition = ResourceOfType(ProcessDefinition,
                                        "process-definition/{id}")

    process_instance = ResourceOfType(ProcessInstance, "process-instance/{id}")
    process_instances = ResourcesOfType(ProcessInstance, "process-instance")

    deployment = ResourceOfType(Deployment, "deployment/{id}")
    deployments = ResourcesOfType(Deployment, "deployment")

    task = ResourceOfType(Task, "task/{id}")
    tasks = ResourcesOfType(Task, "task")
