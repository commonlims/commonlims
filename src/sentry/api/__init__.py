

from .client import ApiClient

client = ApiClient()
