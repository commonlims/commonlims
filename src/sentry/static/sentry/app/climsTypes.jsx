import PropTypes from 'prop-types';
import {Set} from 'immutable';

// Export all auto generated resources:
import {Resources} from 'app/propTypes/resources';

export const Pagination = PropTypes.shape({
  pageLinks: PropTypes.string.isRequired, // The links returned by the backend
  cursor: PropTypes.string.isRequired,
});

// Defines what data is selected in a view, how it's grouped and so on.
// The data itself is expected to by in a dictionary called byIds.
export const ListViewState = PropTypes.shape({
  visibleIds: PropTypes.array.isRequired,
  selectedIds: PropTypes.instanceOf(Set).isRequired,
  groupBy: PropTypes.string,
  search: PropTypes.string.isRequired,
  pagination: Pagination.isRequired,
});

// All lists require this data shape
export const List = {
  byIds: PropTypes.object.isRequired,
  listViewState: ListViewState.isRequired,
};

export const Organization = PropTypes.shape({
  id: PropTypes.string.isRequired,
  slug: PropTypes.string.isRequired,
});

// TODO: all these should have javascript-like property names
export const WorkBatch = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  processDefinitionKey: PropTypes.string.isRequired,
  created_at: PropTypes.string.isRequired, // TODO: Map to date
  updated_at: PropTypes.string.isRequired,
  num_comments: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  subtasks: PropTypes.array.isRequired, // TODO: Describe
  transitions: PropTypes.array.isRequired,
  source: PropTypes.shape({
    substances: PropTypes.array, // TODO: describe
    containers: PropTypes.array,
  }).isRequired,
  target: PropTypes.shape({
    substances: PropTypes.array,
    containers: PropTypes.array,
  }),
  tabs: PropTypes.array.isRequired, // TODO: describe
});

export const ProcessAssignmentStore = PropTypes.object; // TODO!

const ClimsTypes = {
  Organization,
  List,
  WorkBatch,
  ProcessAssignmentStore,
  ...Resources,
};

export default ClimsTypes;
