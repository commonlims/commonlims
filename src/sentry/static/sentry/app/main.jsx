/* global module */
import {hot} from 'react-hot-loader';
import React from 'react';
import {Router, browserHistory} from 'react-router';

import routes from 'app/routes';
import {loadPreferencesState} from 'app/actionCreators/preferences';
import {cache} from 'emotion';
import {CacheProvider} from '@emotion/core';

class Main extends React.Component {
  componentDidMount() {
    loadPreferencesState();
  }

  render() {
    // This step is necessary if you still use css, keyframes or injectGlobal from emotion.
    // Once you remove all the usages of them in your app, you can remove this.
    return (
      <CacheProvider value={cache}>
        <Router history={browserHistory}>{routes()}</Router>;
      </CacheProvider>
    );
  }
}

export default hot(module)(Main);
