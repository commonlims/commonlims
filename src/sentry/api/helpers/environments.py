from __future__ import absolute_import

from sentry.api.exceptions import ResourceDoesNotExist
from sentry.models import Environment


def get_environments(request, organization):
    requested_environments = set(request.GET.getlist('environment'))

    if not requested_environments:
        return []

    environments = list(Environment.objects.filter(
        organization_id=organization.id,
        name__in=requested_environments,
    ))

    if set(requested_environments) != set([e.name for e in environments]):
        raise ResourceDoesNotExist

    return environments
