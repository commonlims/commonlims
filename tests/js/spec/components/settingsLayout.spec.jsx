import React from 'react';
import {render} from 'app-test/helpers/render';
import {shallow} from 'enzyme';

import {Client} from 'app/api';
import SettingsLayout from 'app/views/settings/components/settingsLayout';

describe('SettingsLayout', function () {
  beforeEach(function () {
    Client.clearMockResponses();
    Client.addMockResponse({
      url: '/internal/health/',
      body: {
        problems: [],
      },
    });
    Client.addMockResponse({
      url: '/organizations/',
      body: [TestStubs.Organization()],
    });
    Client.addMockResponse({
      url: '/organizations/org-slug/',
      method: 'DELETE',
      statusCode: 401,
      body: {
        sudoRequired: true,
      },
    });
    Client.addMockResponse({
      url: '/authenticators/',
      body: [],
    });
  });

  it('renders', function () {
    const rendered = render(<SettingsLayout route={{}} routes={[]} />);
    expect(rendered).toMatchSnapshot();
  });

  it('can render navigation', function () {
    const Navigation = () => <div>Navigation</div>;
    const wrapper = shallow(
      <SettingsLayout route={{}} routes={[]} renderNavigation={() => <Navigation />} />
    );

    expect(wrapper.find('Navigation')).toHaveLength(1);
  });
});
