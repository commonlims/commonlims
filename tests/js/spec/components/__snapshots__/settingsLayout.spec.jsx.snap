// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsLayout renders 1`] = `
Array [
  <div
    className="css-18ej6sk-StyledSidebar ea1p9lf0"
  >
    <div
      className="css-1ardvtt-SidebarSectionGroup ea1p9lf1"
    >
      <div
        className="css-1nshi6c-SidebarSectionGroup-SidebarSection ea1p9lf2"
      >
        <div
          className="css-9k9y4m-SidebarDropdownRoot e1fowdfw3"
        >
          <div
            className="css-1bagylm-SidebarDropdownActor e1fowdfw7"
            data-test-id="sidebar-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "outline": "none",
              }
            }
            tabIndex={-1}
          >
            <div
              style={
                Object {
                  "alignItems": "flex-start",
                  "display": "flex",
                }
              }
            >
              <span
                className="avatar e1fowdfw8 css-1n3qnpl-StyledBaseAvatar-StyledAvatar e1z0ohzl0"
                style={
                  Object {
                    "height": "32px",
                    "width": "32px",
                  }
                }
              >
                <svg
                  className="css-3xmr0j-Svg e1knxa8x0"
                  viewBox="0 0 120 120"
                >
                  <rect
                    fill="#315cac"
                    height="120"
                    rx="15"
                    ry="15"
                    width="120"
                    x="0"
                    y="0"
                  />
                  <text
                    fill="#FFFFFF"
                    fontSize="65"
                    style={
                      Object {
                        "dominantBaseline": "central",
                      }
                    }
                    textAnchor="middle"
                    x="50%"
                    y="50%"
                  >
                    FB
                  </text>
                </svg>
              </span>
              <div
                className="css-lg6ezi-OrgAndUserWrapper e1fowdfw4"
              >
                <div
                  className="css-ix2jor-TextOverflow-OrgOrUserName e1fowdfw5"
                >
                  Foo Bar
                   
                  <i
                    className="icon-arrow-down"
                  />
                </div>
                <div
                  className="css-zumxhm-TextOverflow-UserNameOrEmail e1fowdfw6"
                >
                  foo@example.com
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="css-srkfus-SettingsColumn ervyuv13"
  >
    <div
      className="css-1b5ey59-SettingsHeader e16e5sy70"
    >
      <div
        className="css-knfhfe"
      >
        <div
          className="css-1pysja1"
        >
          <div
            className="css-11dln98-Breadcrumbs e18utnk40"
          />
        </div>
        <div
          className="css-7gnmdq-SearchWrapper ed7560h1"
        >
          <div
            className="css-17mgn7r-SearchInputWrapper e146h9au1"
          >
            <svg
              className="e146h9au2 css-a56pm8-StyledSvg-SearchInputIcon e2idor0"
              height="14px"
              viewBox={Object {}}
              width="14px"
            >
              <use
                href="#test"
                xlinkHref="#test"
              />
            </svg>
            <input
              className="css-134g8uj-SearchInput e146h9au3"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              placeholder="Search"
              style={
                Object {
                  "outline": "none",
                }
              }
              tabIndex={-1}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-wm0ppo-MaxWidthContainer ervyuv10"
    >
      <div
        className="css-1qa7uz4-Content ervyuv14"
      />
    </div>
    <footer>
      <div
        className="container"
      >
        <div
          className="pull-right"
        >
          <a
            className="hidden-xs"
            href="/api/"
          >
            API
          </a>
          <a
            href="/docs/"
          >
            Docs
          </a>
          <a
            className="hidden-xs"
            href="https://github.com/commonlims/commonlims"
            rel="noreferrer"
          >
            Contribute
          </a>
        </div>
      </div>
    </footer>
  </div>,
]
`;
